File: python/src/chiptime/metrics/_basics.py
"""Optional analytics on chiptime's honest streams.

Import explicitly (`from chiptime import metrics`) — the core never imports
this module. Everything here inherits the streams' guarantees: sentinels are
already null, zero is real (taxonomy #64), so a 65535 W spike can never
corrupt a curve. Missing data shrinks coverage; it is never filled in.

Names are generic on purpose (no trademarked training-metric names).
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from chiptime.model import Session

MEAN_MAX_MIN_COVERAGE = 0.9  # a window needs >=90% present samples
ZONE_DT_CAP_S = 30.0  # ADR-0005 gap policy reused


def mean_max(values: list[Any], windows: list[int]) -> dict[int, float | None]:
    """Best rolling average per window size, in the RECORD domain.

    At 1 Hz recording (the dominant case) record-domain == time-domain; for
    smart-recording files interpret windows as record counts. Windows with
    less than 90% data coverage return None — absence is not zero.
    """
    nums = [
        float(v) if isinstance(v, (int, float)) and not isinstance(v, bool) else None
        for v in values
    ]
    n = len(nums)
    prefix = [0.0]
    present = [0]
    for v in nums:
        prefix.append(prefix[-1] + (v or 0.0))
        present.append(present[-1] + (1 if v is not None else 0))
    out: dict[int, float | None] = {}
    for w in windows:
        if w <= 0 or w > n:
            out[w] = None
            continue
        best: float | None = None
        min_present = int(w * MEAN_MAX_MIN_COVERAGE + 0.999999)
        for i in range(n - w + 1):
            have = present[i + w] - present[i]
            if have < min_present:
                continue
            avg = (prefix[i + w] - prefix[i]) / have
            if best is None or avg > best:
                best = avg
        out[w] = best
    return out


def time_in_zones(
    times: list[datetime | None], values: list[Any], bounds: list[float]
) -> list[float]:
    """Seconds spent per zone. Zones: (-inf, b0], (b0, b1], ..., (bn, inf) —
    len(bounds)+1 buckets. dt attribution per record, capped at 30 s
    (a gap is a gap, not an hour in zone 2). None samples contribute nowhere."""
    zones = [0.0] * (len(bounds) + 1)
    for i in range(len(times) - 1):
        t0, t1 = times[i], times[i + 1]
        v = values[i]
        if t0 is None or t1 is None or not isinstance(v, (int, float)):
            continue
        dt = (t1 - t0).total_seconds()
        if dt <= 0:
            continue
        dt = min(dt, ZONE_DT_CAP_S)
        z = 0
        for b in bounds:
            if v > b:
                z += 1
            else:
                break
        zones[z] += dt
    return zones


def swolf(session: Session) -> tuple[list[int | None], float | None]:
    """Per-active-length SWOLF (strokes + seconds) and the mean over lengths
    where both parts are present. Pool swimming only (#73)."""
    per: list[int | None] = []
    for ln in session.lengths:
        if ln.length_type != "active":
            continue
        if ln.total_strokes is None or ln.total_elapsed_time_s is None:
            per.append(None)
            continue
        per.append(round(ln.total_strokes + ln.total_elapsed_time_s))
    known = [v for v in per if v is not None]
    return per, (sum(known) / len(known)) if known else None

File: js/src/metrics/basics.ts
/**
 * Optional analytics on chiptime's honest streams.
 *
 * Twin of `python/src/chiptime/metrics/_basics.py`. Everything inherits the
 * streams' guarantees: sentinels are already null, zero is real (taxonomy #64), so
 * a 65535 W spike can never corrupt a curve. Missing data shrinks coverage; it is
 * never filled in.
 */

import type { Session } from "../model.js";
import { pyRound, pySum } from "../numeric.js";

export const MEAN_MAX_MIN_COVERAGE = 0.9; // a window needs >=90% present samples
export const ZONE_DT_CAP_S = 30.0; // ADR-0005 gap policy reused

/**
 * Best rolling average per window size, in the RECORD domain.
 *
 * Windows with less than 90% data coverage return null — absence is not zero.
 */
export function meanMax(
  values: readonly unknown[],
  windows: readonly number[],
): Map<number, number | null> {
  const nums: (number | null)[] = values.map((v) => (typeof v === "number" ? v : null));
  const n = nums.length;
  const prefix: number[] = [0];
  const present: number[] = [0];
  for (const v of nums) {
    prefix.push((prefix[prefix.length - 1] as number) + (v ?? 0));
    present.push((present[present.length - 1] as number) + (v !== null ? 1 : 0));
  }
  const out = new Map<number, number | null>();
  for (const w of windows) {
    if (w <= 0 || w > n) {
      out.set(w, null);
      continue;
    }
    let best: number | null = null;
    const minPresent = Math.trunc(w * MEAN_MAX_MIN_COVERAGE + 0.999999);
    for (let i = 0; i <= n - w; i++) {
      const have = (present[i + w] as number) - (present[i] as number);
      if (have < minPresent) continue;
      const avg = ((prefix[i + w] as number) - (prefix[i] as number)) / have;
      if (best === null || avg > best) best = avg;
    }
    out.set(w, best);
  }
  return out;
}

/**
 * Seconds spent per zone. Zones: (-inf, b0], (b0, b1], ..., (bn, inf) —
 * bounds.length+1 buckets. dt attribution per record, capped at 30 s (a gap is a
 * gap, not an hour in zone 2). Null samples contribute nowhere.
 */
export function timeInZones(
  times: readonly (number | null)[],
  values: readonly unknown[],
  bounds: readonly number[],
): number[] {
  const zones = new Array<number>(bounds.length + 1).fill(0);
  for (let i = 0; i < times.length - 1; i++) {
    const t0 = times[i];
    const t1 = times[i + 1];
    const v = values[i];
    if (t0 === null || t0 === undefined || t1 === null || t1 === undefined) continue;
    if (typeof v !== "number") continue;
    let dt = t1 - t0;
    if (dt <= 0) continue;
    dt = Math.min(dt, ZONE_DT_CAP_S);
    let z = 0;
    for (const b of bounds) {
      if (v > b) z += 1;
      else break;
    }
    zones[z] = (zones[z] as number) + dt;
  }
  return zones;
}

/**
 * Per-active-length SWOLF (strokes + seconds) and the mean over lengths where both
 * parts are present. Pool swimming only (#73).
 */
export function swolf(session: Session): [(number | null)[], number | null] {
  const per: (number | null)[] = [];
  for (const ln of session.lengths) {
    if (ln.lengthType !== "active") continue;
    if (ln.totalStrokes === null || ln.totalElapsedTimeS === null) {
      per.push(null);
      continue;
    }
    per.push(pyRound(ln.totalStrokes + ln.totalElapsedTimeS));
  }
  const known = per.filter((v): v is number => v !== null);
  return [per, known.length > 0 ? pySum(known) / known.length : null];
}
